### PR TITLE
Trusted Flag

### PR DESCRIPTION
**Pull request by jogasser**
- Adds a trusted flag to Gobra, which lets a user assume that the specification of a method or function holds without checking it. To achieve this, it ignores the body of the function or method
- Upgrades sbt to 1.5.7 to fix log4j CVE(see, https://www.scala-lang.org/blog-detail/2021/12/16/state-of-log4j-in-scala-ecosystem.html for more details)

### CODE REVIEW

File: src/main/scala/viper/gobra/frontend/Parser.scala [hunk 1]
```diff
@@ -464,14 +465,21 @@ object Parser {
         "pure" <~ eos ^^^ PureClause
       }
 
-      functSpecClause.* ~ "pure".? ^^ {
+      val specClauses = if(withTrusted) {
+        functSpecClause | "trusted" <~ eos ^^^ TrustedClause
+      } else {
+        functSpecClause
+      }
```
Comment (jcp19) [inline src/main/scala/viper/gobra/frontend/Parser.scala:472]: the variable `specClauses` is unnecessary. We can just add `trusted" <~ eos ^^^ TrustedClause` to the definition of `functSpecClause`.

File: src/main/scala/viper/gobra/frontend/Parser.scala [hunk 2]
```diff
@@ -464,14 +465,21 @@ object Parser {
...
+
+      specClauses.* ~ "pure".? ^^ {
```
Comment (jcp19) [inline src/main/scala/viper/gobra/frontend/Parser.scala:474]: we can revert the changes in this line after incorporating my previous suggestion.

File: src/main/scala/viper/gobra/frontend/Parser.scala [hunk 3]
```diff
@@ -442,19 +442,20 @@ object Parser {
       (idnDef <~ "=") ~ typ ^^ { case left ~ right => PTypeAlias(right, left)}
 
     lazy val functionDecl: Parser[PFunctionDecl] =
-      functionSpec ~ ("func" ~> idnDef) ~ signature ~ specOnlyParser(blockWithBodyParameterInfo) ^^ {
-        case spec ~ name ~ sig ~ body =>
-          PFunctionDecl(name, sig._1, sig._2, spec, body)
+      functionSpec() ~ ("func" ~> idnDef) ~ signature ~ specOnlyParser(blockWithBodyParameterInfo) ^^ {
+        case spec ~ name ~ sig ~ body => PFunctionDecl(name, sig._1, sig._2, spec, body)
+      } | functionSpec(withTrusted = true) ~ ("func" ~> idnDef) ~ signature ~ nestedCurlyBracketsConsumer.? ^^ {
+        case spec ~ name ~ sig ~ _ => PFunctionDecl(name, sig._1, sig._2, spec, None)
```
Comment (jcp19) [inline src/main/scala/viper/gobra/frontend/Parser.scala:448]: I think that there is some repetition here that we can likely avoid. In particular, both branches start by parsing the functionSpec and then obtain the identifier and signature. They only differ in the way that the body is parsed.

File: src/main/scala/viper/gobra/frontend/Parser.scala (hunk 1 above)
Comment (jcp19) [reply src/main/scala/viper/gobra/frontend/Parser.scala:472]: In particular, I do not think that we need the `withTrusted` parameter. Instead, we need to pick the parser for the body according to the result of parsing the spec (look at my suggestion above).

File: src/main/scala/viper/gobra/frontend/Parser.scala (hunk 3 above)
Comment (jcp19) [reply src/main/scala/viper/gobra/frontend/Parser.scala:448]: Something like the following could work:

```scala
    lazy val functionDecl: Parser[PFunctionDecl] =
      for {
        spec <- functionSpec // notice that I removed the param in functionSpec
        name <- ("func" ~> idnDef)
        sig  <- signature
        body <- if (spec.isTrusted) nestedCurlyBracketsConsumer else specOnlyParser(blockWithBodyParameterInfo)
      } yield PFunctionDecl(name, sig._1, sig._2, spec, body)
```

This can likely be improved to look better by using operators from the parsing library instead of the `for` notation

**Review by jcp19 — CHANGES_REQUESTED** [review]

File: src/main/scala/viper/gobra/frontend/Parser.scala [hunk 4]
```diff
@@ -480,8 +488,10 @@ object Parser {
         repsep(expression, ",") ~ ("if" ~> expression).? ^^ PTupleTerminationMeasure
 
     lazy val methodDecl: Parser[PMethodDecl] =
-      functionSpec ~ ("func" ~> receiver) ~ idnDef ~ signature ~ specOnlyParser(blockWithBodyParameterInfo) ^^ {
+      functionSpec() ~ ("func" ~> receiver) ~ idnDef ~ signature ~ specOnlyParser(blockWithBodyParameterInfo) ^^ {
         case spec ~ rcv ~ name ~ sig ~ body => PMethodDecl(name, rcv, sig._1, sig._2, spec, body)
+      } | functionSpec(withTrusted = true) ~ ("func" ~> receiver) ~ idnDef ~ signature ~ nestedCurlyBracketsConsumer.? ^^ {
+        case spec ~ rcv ~ name ~ sig ~ _ => PMethodDecl(name, rcv, sig._1, sig._2, spec, None)
```
Comment (jcp19) [inline src/main/scala/viper/gobra/frontend/Parser.scala:494]: same comments as for `functionDecl`

File: src/main/scala/viper/gobra/frontend/Parser.scala (hunk 3 above)
Comment (jogasser) [reply src/main/scala/viper/gobra/frontend/Parser.scala:448]: This was actually my first attempt of solving this task. I used the `into` operator of the parsing library. The problem with an approach like this is, that if you parse the function spec first, the position of the `PFunctionDecl` node moves down to the `func` key word, implying that the spec doesn't belong to a function declaration. I discussed this with Linard and we came to the conclusion, that this isn't desired behaviour. This is the reason why I used this non-elegant approach

Comment (jogasser) [reply src/main/scala/viper/gobra/frontend/Parser.scala:448]: `regressions/features/interfaces/proofGenFail1.gobra` would be a test case, where the above described behaviour would cause problems

Comment (jcp19) [reply src/main/scala/viper/gobra/frontend/Parser.scala:448]: I am not sure if I understand what you mean. Do you mean that the position associated with this AST node is not the intended one? If so, isn't this something that we can overcome by using the `at` method to specify exactly the position of a node? (there are some examples of this method in the Parser class)

Comment (jcp19) [reply src/main/scala/viper/gobra/frontend/Parser.scala:448]: PS: verifying the file `regressions/features/interfaces/proofGenFail1.gobra` with my proposed code seems to lead to the same exact output as the latest commit of this branch. Is it possible that the erroneous behaviour that you found is caused by some special behaviour of the `into` operator?

Comment (jogasser) [reply src/main/scala/viper/gobra/frontend/Parser.scala:448]: I didn't know about the `at` operator, this fixes the issue, thanks!

About the problem, the output for the test case is almost the same, the test output should be: 
```
...
[info] 	<proofGenFail1.gobra:24:1> Generated implementation proof (*counter implements interface{ hasNext, next }) failed. Precondition of call to implementation method might not hold. 
[info] Permission to x might not suffice.
[info] 	<proofGenFail1.gobra:30:1> Generated implementation proof (*counter implements interface{ hasNext, next }) failed. Precondition of call to implementation method might not hold. 
[info] Permission to x.f might not suffice.
...
```

without the `at` operator, even with your code, the output looks something like this (there is a difference in the error positions): 
```
...
[info] 	<proofGenFail1.gobra:25:39> Generated implementation proof (*counter implements interface{ hasNext, next }) failed. Precondition of call to implementation method might not hold. 
[info] Permission to x might not suffice.
[info] 	<proofGenFail1.gobra:33:42> Generated implementation proof (*counter implements interface{ hasNext, next }) failed. Precondition of call to implementation method might not hold. 
[info] Permission to x.f might not suffice.
...
```

**Review by jcp19 — APPROVED** [review]
It looks good to me now. @ArquintL do you have other comments? Otherwise, I suggest that we merge this when the CI has gone through.

File: docs/tutorial.md [hunk 5]
```diff
@@ -44,6 +44,7 @@ import (
 ```go
 requires ... // preconditions 
 ensures  ... // postconditions 
+trusted      // trusted flag (optional)
```
Comment (ArquintL) [inline docs/tutorial.md:47]: thanks for keeping the tutorial up-to-date!

File: src/main/scala/viper/gobra/frontend/Parser.scala [hunk 6]
```diff
@@ -442,36 +442,39 @@ object Parser {
...
-      functionSpec ~ ("func" ~> idnDef) ~ signature ~ specOnlyParser(blockWithBodyParameterInfo) ^^ {
-        case spec ~ name ~ sig ~ body =>
-          PFunctionDecl(name, sig._1, sig._2, spec, body)
-      }
+      for {
+        spec <- functionSpec
+        name <- "func" ~> idnDef
+        sig  <- signature
+        body <- if (spec.isTrusted) nestedCurlyBracketsConsumer else specOnlyParser(blockWithBodyParameterInfo)
+      } yield PFunctionDecl(name, sig._1, sig._2, spec, body).at(spec)
```
Comment (ArquintL) [inline src/main/scala/viper/gobra/frontend/Parser.scala:450]: Is it necessary to  call `.at(spec)` here? I would expect that Kiama sets the positional information itself but I'm not sure in the context of `for ... yield`. Letting Kiama do it would have the advantage that the end position is presumably after the body. As you use `spec` right now, the function's position only spans pre- & postconditions

**Review by ArquintL — APPROVED** [review]
One small question regarding positional information, otherwise looks good to me

File: src/main/scala/viper/gobra/frontend/Parser.scala (hunk 6 above)
Comment (jogasser) [reply src/main/scala/viper/gobra/frontend/Parser.scala]: You are correct, I replaced `.at(...)` with `.range(...)`. 

I don't really know if it makes more sense to let Kiama handle the position, since, with a solution where the spec gets parsed before the body, Kiama wouldn't include the spec in the function position. 

The previous solution, where I let Kiama set the position correctly(see `src/main/scala/viper/gobra/frontend/Parser.scala` in [62cdff1](https://github.com/viperproject/gobra/pull/378/commits/62cdff16b205bda2713f456a2d313215f5d25845)), is, as joao said a little bit repetitive and unelegant.

**Comment by ArquintL** [comment]
Thanks for the improvement. However, this new positions do not cover the body if the trusted flag is set. I've created the attached patch that addresses this. You should simply be able to run `git apply parser.txt` (it's a patch file but GitHub does not allow me to upload files with that extension ^^) and then commit the changes (I can't do it as I don't have access to your fork).

[parser.txt](https://github.com/viperproject/gobra/files/7813625/parser.txt)

If it should ever be handy, I've used the following snippet to print the positional information of functions in the Success case in `parseSources`:
```
val fnDecls = ast.declarations.collect {
  case n: PFunctionDecl => n
}

println(s"position for function decls:\n${fnDecls.map(fn => s"${fn.id.name}: ${pom.positions.getStart(fn).map(pos => s"${pos.line}:${pos.column}")} - ${pom.positions.getFinish(fn).map(pos => s"${pos.line}:${pos.column}")}").mkString("\n")}")
```

**Review by ArquintL — CHANGES_REQUESTED** [review]
Please apply my patch first

**Review by ArquintL — APPROVED** [review]

**Review by jcp19 — APPROVED** [review]
LGTM
